### PR TITLE
[HttpFoundation] [Request] Throw dedicated exceptions.

### PR DESCRIPTION
**Pull request by SpacePossum**
| Q             | A
| ------------- | ---
| Branch?       | master (but the lower the better ;) )
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | not really
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

I'm trying to deal with invalid requests being sent to one of my sites.
While this is a Silex site I think this PR will help dealing with such requests and will be useful for SF developers as well.

The problem is that sometimes a request is send with host data like "my.test.com/".
The trailing slash will cause the Request class to throw an exception when asking for the host (for example through `getHost()`).

In order to deal with these kind of errors I setup something like:

```php
// bootstrap.php

$app = new Application(); // this is a Silex app

// more bootstrap stuff

$app->register(new MonologServiceProvider(), [/**/]);
$app['monolog.listener'] = function () use ($app) {
    return new class($app['logger'], $app['folders.log']) extends LogListener {
        protected function logException(\Exception $e)
        {
            if ($e instanceof HttpExceptionInterface) {
                if ($e->getStatusCode() >= 500) { // don't log 400 errors
                    $this->logger->critical($message, ['exception' => $e]);
                }
            } else {
                // FIXME deal with host problem here
                $this->logger->critical($message, ['exception' => $e]);
            }		
        }
    }
};

$app->error(function (\Exception $e, Request $request, $code = null) use ($app) {
    // FIXME deal with host problem here
}

// more bootstrap stuff

$app->run();
```

Dealing with the exception at these points is hard because the Request class throws an SPL `\UnexpectedValueException` exception.
At the point of handling it is hard to see this exception comes from the Request class and not from somewhere else in the system (for example a controller dealing with the Request object).

It is not possible to deal with exception within the controllers and/or other part of the application on user level.
This is because the exception is thrown _before_ the Request is given to a controller.

Example trace:
```bash
#0 /var/www/digital-shop-intermediate-site/vendor/symfony/http-foundation/Request.php(1032): Symfony\Component\HttpFoundation\Request->getHost()
#1 /var/www/digital-shop-intermediate-site/vendor/symfony/http-foundation/Request.php(1062): Symfony\Component\HttpFoundation\Request->getHttpHost()
#2 /var/www/digital-shop-intermediate-site/vendor/symfony/http-foundation/Request.php(1716): Symfony\Component\HttpFoundation\Request->getSchemeAndHttpHost()
#3 /var/www/digital-shop-intermediate-site/vendor/symfony/http-foundation/Request.php(1046): Symfony\Component\HttpFoundation\Request->prepareRequestUri()
#4 /var/www/digital-shop-intermediate-site/vendor/symfony/http-foundation/Request.php(1768): Symfony\Component\HttpFoundation\Request->getRequestUri()
#5 /var/www/digital-shop-intermediate-site/vendor/symfony/http-foundation/Request.php(924): Symfony\Component\HttpFoundation\Request->prepareBaseUrl()
#6 /var/www/digital-shop-intermediate-site/vendor/symfony/routing/RequestContext.php(73): Symfony\Component\HttpFoundation\Request->getBaseUrl()
#7 /var/www/digital-shop-intermediate-site/vendor/symfony/http-kernel/EventListener/RouterListener.php(71): Symfony\Component\Routing\RequestContext->fromRequest(Object(Symfony\Component\HttpFoundation\Request))
#8 /var/www/digital-shop-intermediate-site/vendor/symfony/http-kernel/EventListener/RouterListener.php(90): Symfony\Component\HttpKernel\EventListener\RouterListener->setCurrentRequest(Object(Symfony\Component\HttpFoundation\Request))
#9 [internal function]: Symfony\Component\HttpKernel\EventListener\RouterListener->onKernelRequest(Object(Symfony\Component\HttpKernel\Event\GetResponseEvent), 'kernel.request', Object(Symfony\Component\EventDispatcher\EventDispatcher))
#10 /var/www/digital-shop-intermediate-site/vendor/symfony/event-dispatcher/EventDispatcher.php(174): call_user_func(Array, Object(Symfony\Component\HttpKernel\Event\GetResponseEvent), 'kernel.request', Object(Symfony\Component\EventDispatcher\EventDispatcher))
#11 /var/www/digital-shop-intermediate-site/vendor/symfony/event-dispatcher/EventDispatcher.php(43): Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(Array, 'kernel.request', Object(Symfony\Component\HttpKernel\Event\GetResponseEvent))
#12 /var/www/digital-shop-intermediate-site/vendor/symfony/http-kernel/HttpKernel.php(129): Symfony\Component\EventDispatcher\EventDispatcher->dispatch('kernel.request', Object(Symfony\Component\HttpKernel\Event\GetResponseEvent))
#13 /var/www/digital-shop-intermediate-site/vendor/symfony/http-kernel/HttpKernel.php(68): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#14 /var/www/digital-shop-intermediate-site/vendor/silex/silex/src/Silex/Application.php(496): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true)
#15 /var/www/digital-shop-intermediate-site/vendor/silex/silex/src/Silex/Application.php(477): Silex\Application->handle(Object(Symfony\Component\HttpFoundation\Request))
#16 /var/www/digital-shop-intermediate-site/web/index.php(13): Silex\Application->run()
```

It will not only be a lot of work modifying all components dealing with the host of a Request (either by `getHost` or `getHttpHost`) before the Request is routed to a controller, it will also be hard to make sure when new components are added (or existing modified) deal with the exception correctly.

Therefor I propose to let the Request class throw exceptions of meaningful (/dedicated) classes. That way people can deal with these exception as they want.

### CODE REVIEW

**Comment by SpacePossum** [comment]
As a follow up on the invalid host, the request is from:
```
HTTP_USER_AGENT": "Mozilla\/5.0 (compatible; Googlebot\/2.1; +http:\/\/www.google.com\/bot.html)
```

So somehow at least google thinks using a host with a trailing slash is valid.
Is a host like `go.test.com/` really invalid?

File: src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation\Exception;
```
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php]: In other components, exceptions implement an interface specific to the component.

Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php]: Which interface would be appropriate in this case?
For example currently the class throws exceptions of the class [ConflictingHeadersException]( https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Exception/ConflictingHeadersException.php) but that doesn't implement any interface.

**Comment by SpacePossum** [comment]
reworked, @Ener-Getick please have a look :)

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 2]
```diff
@@ -1220,7 +1226,7 @@ public function getHost()
         // check that it does not contain forbidden characters (see RFC 952 and RFC 2181)
         // use preg_replace() instead of preg_match() to prevent DoS attacks with long host names
         if ($host && '' !== preg_replace('/(?:^\[)?[a-zA-Z0-9-:\]_]+\.?/', '', $host)) {
-            throw new \UnexpectedValueException(sprintf('Invalid Host "%s"', $host));
+            throw new InvalidHostException($host, sprintf('Invalid host "%s".', $host));
```
Comment (chalasr) [inline src/Symfony/Component/HttpFoundation/Request.php]: Couldn't we just pass the message as it contains the host, such as done for `UntrustedHostException`?

Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Request.php]: I would like to get the `host` that triggered the exception to be thrown when I'm dealing with the exception. Relying on the message (format) of an exception is never a good idea IMHO.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 3]
```diff
@@ -620,12 +623,12 @@ public static function setTrustedHeaderName($key, $value)
...
      * @return string The header name
      *
-     * @throws \InvalidArgumentException
+     * @throws InvalidTrustedHeaderException
      */
     public static function getTrustedHeaderName($key)
     {
         if (!array_key_exists($key, self::$trustedHeaders)) {
-            throw new \InvalidArgumentException(sprintf('Unable to get the trusted header name for key "%s".', $key));
+            throw new InvalidTrustedHeaderException($key, null, sprintf('Unable to get the trusted header name for key "%s".', $key));
```
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Request.php]: Isn't an `InvalidArgumentException` localized in the component enough and more appropriate?

File: src/Symfony/Component/HttpFoundation/Tests/RequestTest.php [hunk 4]
```diff
@@ -1936,7 +1937,7 @@ public function testHostValidity($host, $isValid, $expectedHost = null, $expecte
                 $this->assertSame($expectedPort, $request->getPort());
             }
         } else {
-            $this->setExpectedException('UnexpectedValueException', 'Invalid Host');
+            $this->setExpectedException('Symfony\Component\HttpFoundation\Exception\InvalidHostException', 'Invalid host');
```
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Tests/RequestTest.php]: You can use the short syntax here `Exception::class`

File: src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * The HTTP request contains invalid host data.
+ *
+ * @author SpacePossum
+ */
+final class InvalidHostException extends AbstractHostException
```
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php]: I think this one should be internal.

File: src/Symfony/Component/HttpFoundation/Tests/RequestTest.php (hunk 4 above)
Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Tests/RequestTest.php]: ah SF upped requirements to PHP 5.5, good point will update

File: src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php (hunk 5 above)
Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php]: :+1: agreed

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 3 above)
Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Request.php]: Happy to revert :)

File: src/Symfony/Component/HttpFoundation/Exception/AbstractHostException.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    /**
+     * @var string
+     */
+    private $host;
+
+    /**
+     * @param string $host
+     * @param string $message
+     */
+    public function __construct($host, $message)
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/Exception/AbstractHostException.php]: Why not making `$message` optional (if null, default to `Invalid host "'.$host.'"`), so you can simply throw `new InvalidHostException($host)`.

File: src/Symfony/Component/HttpFoundation/Exception/InvalidTrustedHeaderException.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+     * @var string
+     */
+    private $value;
+
+    /**
+     * @param string      $header
+     * @param string|null $value
+     * @param string      $message
+     */
+    public function __construct($header, $value, $message)
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidTrustedHeaderException.php]: Same.. make message optional?

File: src/Symfony/Component/HttpFoundation/Exception/InvalidTrustedHeaderException.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+{
+    /**
+     * @var string
+     */
+    private $header;
+
+    /**
+     * @var string
+     */
+    private $value;
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidTrustedHeaderException.php]: Having `getHeader() / getValue()` would be nice? Right now these are basically useless.

File: src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php (hunk 5 above)
Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php]: On second thought, I don't think so, people want to deal with this exception by catching it. Marking it as internal will cause IDE's to complain about using an internal class.

File: src/Symfony/Component/HttpFoundation/Exception/AbstractHostException.php (hunk 6 above)
Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Exception/AbstractHostException.php]: Making it `null` able will break with the design of the base exception class itself, which is IIRC something people don't want in the SF project.
Besides there is no way this class can generate a useful/correct message based on the value of `$host` alone, the logic that figured out the `$host` value is incorrect and throws the exception knows what is wrong and can therefor set a meaningful message only.

File: src/Symfony/Component/HttpFoundation/Exception/InvalidTrustedHeaderException.php (hunk 8 above)
Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidTrustedHeaderException.php]: thank, indeed, however as opt [here](https://github.com/symfony/symfony/pull/20389#discussion_r86617704) I will remove the class all together.

File: src/Symfony/Component/HttpFoundation/Exception/AbstractHostException.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * THe HTTP request contains incorrect host data.
+ *
+ * @author SpacePossum
+ */
+abstract class AbstractHostException extends \UnexpectedValueException implements ExceptionInterface
+{
+    /**
+     * @var string
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/Exception/AbstractHostException.php]: I believe unnecessary docs can be removed.. same for the getter.

**Comment by hhamon** [comment]
If this PR is accepted, it will only go into Symfony 4.0 only for BC reasons I suppose. You should also trigger some deprecation errors before every new exception types is thrown.

**Comment by ogizanagi** [comment]
@hhamon : Why ? There is no BC break, as the new exceptions extends `\UnexpectedValueException` which was thrown before, right ?

File: src/Symfony/Component/HttpFoundation/Tests/RequestTest.php [hunk 10]
```diff
@@ -1872,8 +1874,8 @@ public function testTrustedHosts()
         try {
             $request->getHost();
             $this->fail('Request::getHost() should throw an exception when host is not trusted.');
-        } catch (\UnexpectedValueException $e) {
-            $this->assertEquals('Untrusted Host "evil.com"', $e->getMessage());
+        } catch (UntrustedHostException $e) {
+            $this->assertEquals('Untrusted host "evil.com".', $e->getMessage());
```
Comment (ogizanagi) [inline src/Symfony/Component/HttpFoundation/Tests/RequestTest.php]: I think `host` was capitalized referencing the Host header. I'm not sure this has to be changed (same for other occurrences).

**Comment by SpacePossum** [comment]
>If this PR is accepted, it will only go into Symfony 4.0 only for BC reasons I suppose. 

I agree with @ogizanagi here, there is no real BC break. The only way this can break is people rely on the exception being an instance of `\UnexpectedValueException` but not inherits from it.

>You should also trigger some deprecation errors before every new exception types is thrown.

no, the logic of the exception being thrown is not deprecated and so it should not trigger such warnings.

File: src/Symfony/Component/HttpFoundation/Tests/RequestTest.php (hunk 10 above)
Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Tests/RequestTest.php]: alright :+1:   updated back to `Host`

File: src/Symfony/Component/HttpFoundation/Exception/AbstractHostException.php (hunk 9 above)
Comment (dunglas) [inline src/Symfony/Component/HttpFoundation/Exception/AbstractHostException.php]: They are useless, they can be inferred from the constructor's doc and most tools support this.

**Comment by dunglas** [comment]
👍

File: src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php (hunk 5 above)
Comment (GuilhemN) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php]: Indeed sorry I didn't commented on the right class... I was thinking about `AbstractHostException`.

Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Exception/InvalidHostException.php]: :+1:  I've updated that class.

**Comment by SpacePossum** [comment]
@dunglas I've removed the PHPDoc's as requested, anything else you want me to change to get the PR into the status that is ready and marked as reviewed?

**Comment by dunglas** [comment]
It's ok for me, you now need to wait that at least another core team member review it and agree to merge it.

**Comment by xabbuh** [comment]
👍 for me (but for 3.3)

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 11]
```diff
@@ -1940,7 +1945,12 @@ private static function createRequestFromFactory(array $query = array(), array $
             $request = call_user_func(self::$requestFactory, $query, $request, $attributes, $cookies, $files, $server, $content);
 
             if (!$request instanceof self) {
-                throw new \LogicException('The Request factory must return an instance of Symfony\Component\HttpFoundation\Request.');
+                throw new \UnexpectedValueException(
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/Request.php]: BC break :) although this is the better exception :(

Comment (SpacePossum) [inline src/Symfony/Component/HttpFoundation/Request.php]: reverted, I don't care to much about that one anyway so ;)

**Comment by SpacePossum** [comment]
> but for 3.3

does this mean I've to retarget this PR?

**Comment by fabpot** [comment]
Closing in favor of #20962

**Pull request closed without merging**